File: organisation/organisation-model/src/main/java/se/mithlond/services/organisation/model/address/CategorizedAddress.java
/*
 * #%L
 * Nazgul Project: mithlond-services-organisation-model
 * %%
 * Copyright (C) 2015 Mithlond
 * %%
 * Licensed under the jGuru Europe AB license (the "License"), based
 * on Apache License, Version 2.0; you may not use this file except
 * in compliance with the License.
 * 
 * You may obtain a copy of the License at
 * 
 *       http://www.jguru.se/licenses/jguruCorporateSourceLicense-2.0.txt
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package se.mithlond.services.organisation.model.address;

import se.jguru.nazgul.tools.validation.api.exception.InternalStateValidationException;
import se.mithlond.services.organisation.model.Category;
import se.mithlond.services.organisation.model.CategoryProducer;
import se.mithlond.services.organisation.model.Listable;
import se.mithlond.services.organisation.model.Organisation;
import se.mithlond.services.organisation.model.OrganisationPatterns;
import se.mithlond.services.shared.spi.algorithms.Validate;

import javax.persistence.CascadeType;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.ManyToOne;
import javax.persistence.NamedQueries;
import javax.persistence.NamedQuery;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlIDREF;
import javax.xml.bind.annotation.XmlType;

/**
 * A classified and categorized Address that belongs to a certain Organisation.
 * CategorizedAddresses should be available only within service calls from the Organisation to which they belong.
 *
 * @author <a href="mailto:lj@jguru.se">Lennart J&ouml;relid</a>, jGuru Europe AB
 */
@NamedQueries({
        @NamedQuery(name = CategorizedAddress.NAMEDQ_GET_BY_ORGANISATION_ID_AND_CATEGORY_IDS,
                query = "select a from CategorizedAddress a "
                        + "where a.owningOrganisation.id = :" + OrganisationPatterns.PARAM_ORGANISATION_ID
                        + " and ( 0 = :" + OrganisationPatterns.PARAM_NUM_CATEGORYIDS
                        + " or a.category.categoryID in :" + OrganisationPatterns.PARAM_CATEGORY_IDS + " ) "
                        + " order by a.shortDesc"),
        @NamedQuery(name = "CategorizedAddress.getByOrganisationAndClassification",
                query = "select a from CategorizedAddress a "
                        + "where a.owningOrganisation.organisationName like :organisationName "
                        + "and a.category.classification like :classification order by a.shortDesc"),
        @NamedQuery(name = "CategorizedAddress.getByOrganisationAndShortDesc",
                query = "select a from CategorizedAddress a "
                        + "where a.owningOrganisation.organisationName like :organisationName "
                        + "and a.shortDesc like :shortDesc order by a.shortDesc"),
        @NamedQuery(name = CategorizedAddress.NAMEDQ_GET_BY_SEARCHPARAMETERS,
                query = "select a from CategorizedAddress a "
                        + "where a.fullDesc like :" + OrganisationPatterns.PARAM_FULL_DESC
                        + " and a.shortDesc like :" + OrganisationPatterns.PARAM_SHORT_DESC
                        + " and ( :" + OrganisationPatterns.PARAM_CLASSIFICATION_IDS + " > 0 "
                        + "      and a.category.classification in :" + OrganisationPatterns.PARAM_CATEGORY_IDS + " ) "
                        + " and ( :" + OrganisationPatterns.PARAM_NUM_ORGANISATIONIDS + " > 0 "
                        + "      and a.owningOrganisation.id in :" + OrganisationPatterns.PARAM_ORGANISATION_IDS + " ) "
                        + " and a.address.careOfLine like :" + OrganisationPatterns.PARAM_ADDRESSCAREOFLINE
                        + " and a.address.city like :" + OrganisationPatterns.PARAM_CITY
                        + " and a.address.country like :" + OrganisationPatterns.PARAM_COUNTRY
                        + " and a.address.departmentName like :" + OrganisationPatterns.PARAM_DEPARTMENT
                        + " and a.address.description like :" + OrganisationPatterns.PARAM_DESCRIPTION
                        + " and a.address.number like :" + OrganisationPatterns.PARAM_NUMBER
                        + " and a.address.street like :" + OrganisationPatterns.PARAM_STREET
                        + " and a.address.zipCode like :" + OrganisationPatterns.PARAM_ZIPCODE
                        + " order by a.category.classification, a.shortDesc")

})
@Entity
@XmlType(namespace = OrganisationPatterns.NAMESPACE, propOrder = {"address", "category"})
@XmlAccessorType(XmlAccessType.FIELD)
public class CategorizedAddress extends Listable implements Comparable<CategorizedAddress> {

    /**
     * NamedQuery for getting CategorizedAddresses by information defined within a search parameters.
     */
    public static final String NAMEDQ_GET_BY_SEARCHPARAMETERS =
            "CategorizedAddress.getBySearchParams";

    /**
     * NamedQuery for getting CategorizedAddresses by organisation and category ID.
     */
    public static final String NAMEDQ_GET_BY_ORGANISATION_ID_AND_CATEGORY_IDS =
            "CategorizedAddress.getByOrganisationIdAndCategoryIDs";

    /**
     * The constant value indicating that a {@link CategorizedAddress} is for Activities.
     */
    public static final String ACTIVITY_CLASSIFICATION = "activity_locale";

    // Internal state
    @ManyToOne(optional = false, cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH})
    @XmlIDREF
    @XmlAttribute(required = true, name = "categoryReference")
    private Category category;

    @Embedded
    @XmlElement(required = true, nillable = false)
    private Address address;

    /**
     * JAXB/JPA-friendly constructor.
     */
    public CategorizedAddress() {
    }

    /**
     * Convenience constructor, creating a CategorizedAddress with a single category.
     *
     * @param shortDesc    The mandatory and non-empty short description of this Listable entity.
     * @param fullDesc     The full description of this entity (up to 2048 chars), visible
     *                     in detailed listings. May not be null or empty.
     * @param organisation The organisation within which this Listable exists.
     * @param address      The address to categorize. Not null.
     * @param producer     The CategoryProducer used to acquire a Category.
     */
    public CategorizedAddress(final String shortDesc,
                              final String fullDesc,
                              final CategoryProducer producer,
                              final Organisation organisation,
                              final Address address) {

        // Delegate
        super(shortDesc, fullDesc, organisation);

        // Check sanity
        Validate.notNull(producer, "producer");

        // Assign internal state
        this.address = address;
        this.category = producer.getCategory();
    }

    /**
     * Assigns the JPA ID of this CategorizedAddress.
     *
     * @param id the JPA ID of this CategorizedAddress.
     */
    @Override
    public void setId(final long id) {
        super.setId(id);
    }

    /**
     * @return The Category of this CategorizedAddress. Not null or empty.
     */
    public Category getCategory() {
        return category;
    }

    /**
     * @return The Address of this CategorizedAddress.
     */
    public Address getAddress() {
        return address;
    }

    /**
     * Assigns the supplied Address to this CategorizedAddress.
     *
     * @param address The address to categorize. Not null.
     */
    public void setAddress(final Address address) {

        // Check sanity
        Validate.notNull(address, "address");

        // Assign internal state
        this.address = address;
    }

    /**
     * Assigns the supplied Category to this CategorizedAddress.
     *
     * @param category The single category of this CategorizedAddress. Not null.
     */
    public void setCategory(final Category category) {

        // Check sanity
        Validate.notNull(category, "Cannot handle null category argument.");

        // Assign internal state
        this.category = category;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int compareTo(final CategorizedAddress that) {

        // Check sanity
        if (that == null) {
            return -1;
        }
        if (that == this) {
            return 0;
        }

        int toReturn = getAddress().compareTo(that.getAddress());
        if (toReturn == 0) {
            toReturn = category.compareTo(that.getCategory());
        }

        // All done.
        return toReturn;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean equals(final Object that) {

        // Check sanity
        if (this == that) {
            return true;
        }
        if (!(that instanceof CategorizedAddress)) {
            return false;
        }

        // Delegate
        return this.hashCode() == that.hashCode();
    }

    /**
     * The hashCode implementation only considers the hashCode of the Address and each Category.
     * <p/>
     * {@inheritDoc}
     */
    @Override
    public int hashCode() {
        return 31 * address.hashCode() + category.hashCode();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return "CategorizedAddress [" + getCategory() + ": " + getAddress() + "]";
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void validateListableEntityState() throws InternalStateValidationException {

        InternalStateValidationException.create()
                .notNull(address, "address")
                .notNull(category, "category")
                .endExpressionAndValidate();
    }
}

File: organisation/organisation-domain-model/src/main/kotlin/se/mithlond/services/organisation/domain/model/address/CategorizedAddress.kt
/*-
 * #%L
 * Nazgul Project: mithlond-services-organisation-domain-model
 * %%
 * Copyright (C) 2018 jGuru Europe AB
 * %%
 * Licensed under the jGuru Europe AB license (the "License"), based
 * on Apache License, Version 2.0; you may not use this file except
 * in compliance with the License.
 * 
 * You may obtain a copy of the License at
 * 
 *       http://www.jguru.se/licenses/jguruCorporateSourceLicense-2.0.txt
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package se.mithlond.services.organisation.domain.model.address

import se.mithlond.services.organisation.domain.model.Category
import se.mithlond.services.organisation.domain.model.Listable
import se.mithlond.services.organisation.domain.model.Organisation
import java.io.Serializable
import javax.persistence.Access
import javax.persistence.AccessType
import javax.persistence.Basic
import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Embedded
import javax.persistence.Entity
import javax.persistence.ForeignKey
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.SequenceGenerator
import javax.persistence.Table
import javax.persistence.UniqueConstraint

/**
 * A classified and categorized Address that belongs to a certain Organisation.
 * CategorizedAddresses should be available only within service calls from the
 * Organisation to which they belong.
 *
 * @param id The JPA ID of this Domain Entity.
 * @param shortDesc The mandatory and non-empty short description of this Listable entity.
 * Typically used within short info boxes and pop-ups.
 * @param fullDesc The full description of this Listable entity, visible in detailed listings.
 * May not be null or empty. Typically used within longer info boxes or modal description displays.
 * @param owningOrganisation The [Organisation] owning this [CategorizedAddress].
 * @param category the [Category] of this [CategorizedAddress]
 * @param address the [Address] within this [CategorizedAddress]
 *
 * @author [Lennart J&ouml;relid](mailto:lj@jguru.se), jGuru Europe AB
 * @see Listable
 */
@Entity
@Access(AccessType.FIELD)
@Table(schema = "organisations", uniqueConstraints = [
    UniqueConstraint(
            name = "unq_shortdesc_per_category_org",
            columnNames = ["shortDesc", "category_id", "owningOrganisation_id"])
])
data class CategorizedAddress(

        @field:Id
        @field:GeneratedValue(strategy = GenerationType.IDENTITY, generator = "seq_CategorizedAddress")
        @field:SequenceGenerator(schema = "organisations", name = "seq_CategorizedAddress",
                sequenceName = "seq_CategorizedAddress", allocationSize = 1)
        @field:Column(name = "id", updatable = false, nullable = false)
        var id: Long? = null,

        @field:Basic(optional = false)
        @field:Column(nullable = false, length = 254)
        override var shortDesc: String,

        @field:Basic(optional = false)
        @field:Column(nullable = false, length = 2048)
        override var fullDesc: String,

        @field:ManyToOne(optional = false)
        @field:JoinColumn(
                name = "organisation_id",
                nullable = false,
                foreignKey = ForeignKey(name = "fk_cataddress_organisation"))
        override var owningOrganisation: Organisation,

        @field:ManyToOne(optional = false, cascade = [CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH])
        @field:JoinColumn(
                name = "category_id",
                nullable = false,
                foreignKey = ForeignKey(name = "fk_cataddress_category"))
        var category: Category,

        @Embedded
        var address: Address

) : Listable, Serializable, Comparable<CategorizedAddress> {

    override fun compareTo(other: CategorizedAddress): Int {

        var toReturn = this.owningOrganisation.compareTo(other.owningOrganisation)

        if(toReturn == 0) {
            toReturn = this.category.compareTo(other.category)
        }

        if(toReturn == 0) {
            toReturn = this.address.compareTo(other.address)
        }

        // All Done.
        return toReturn
    }

    companion object {

        /**
         * The constant value indicating that a [CategorizedAddress] is for Activities.
         */
        @JvmStatic
        val ACTIVITY_CLASSIFICATION = "activity_locale"
    }
}
